File: python/leetcode/advenced-algorithm/sliding-window-maximum.py
from typing import List
from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        deq, result = deque(), []

        for i, num in enumerate(nums):
            while deq and nums[deq[-1]] < num:
                deq.pop()

            deq.append(i)

            if deq[0] == i - k:
                deq.popleft()

            if i >= k - 1:
                result.append(nums[deq[0]])

        return result


print(Solution().maxSlidingWindow([1, 3, -1, -3, 5, 3, 6, 7], 3))

File: javascript/leetcode/advanced-algorithm/sliding-window-maximum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function (nums, k) {
  const current = [];
  const result = [];

  for (let i = 0; i < nums.length; i++) {
    while (current && nums[current.at(-1)] < nums[i]) {
      current.pop();
    }

    current.push(i);

    if (current[0] === i - k) {
      current.shift();
    }

    if (i >= k - 1) {
      result.push(nums[current[0]]);
    }
  }

  return result;
};

console.log(maxSlidingWindow([1, 3, -1, -3, 5, 3, 6, 7], 3));
